### PR TITLE
added handling BlockedConnection, especially required for dealing with o...

### PR DESCRIPTION
**Pull request by Morgaroth**
...verflowed queue and slow consumer, Connecton Actor may be subscribed on blocking connection events and, when he is, broadcasts information about blocked connection to channel actors. Channel actors handle publishing messages in if where is checked state od connection and message is published or no, and sender is informed about it (be replying result to sender())

Maintainers, I'm asking You for quick look at changes and some feedback.

### CODE REVIEW

**Comment by Morgaroth** [comment]
Are those tests deterministic? :)
On my own all pass :P

**Comment by Morgaroth** [comment]
One thing I miss, some examples in readme should be changed, if repository maintainers agree this PR I change those exaples too

**Comment by reinierl** [comment]
This brings some interesting features, of course the handling of the blocking notifications, but also the feedback by the ChannelActor about the success / failure of the ChannelMessage.

Still I see one problem. That is that the ChannelActor will queue up messages in memory when the connection is lost, but will report failure back to the sender when the connection is blocked. I don't see why this distinction would be there. I think it is better to queue up messages in memory also for blocked connections.

**Comment by Morgaroth** [comment]
Queing messages when connection is blocked, good idea.

So may I explain my motives, main idea is to block incoming traffic when machine resources ends, in that case application (this one using rabbit) should stop accepting or limit incomming data (in case of http application by returning 4xx or 5xx), for survive and notify humans about this situation. In this case it is ensured that accepted data will be handled in server. 
But, when messages are queued in memory, in situation of more incoming data than application is able to handle, unhandled messages will be queued in memory and in the end jvm throw out of memory error, jvm gone down and all messages waiting for queing will be lost, but remote application will think that everything is okay, but not in real.

What You think about this case, case when application is poisoning by remote, queing messages in memory  and is going to out of memory ?

Sending confirmation messages to publisher actor should be delayed to moment of queuing to rabbit in real ?

File: src/main/scala/com/thenewmotion/akka/rabbitmq/ConnectionActor.scala [hunk 1]
```diff
@@ -80,20 +84,30 @@ class ConnectionActor(
           goto(Disconnected) using NoConnection replying ChannelCreated(child)
       }
 
-    case Event(AmqpShutdownSignal(cause), Connected(connection)) =>
+    case Event(AmqpShutdownSignal(cause), Connected(connection, _)) =>
       if (!cause.isInitiatedByApplication) reconnect(connection)
       goto(Disconnected) using NoConnection
+
+    case Event(blocked: QueueBlocked, data: Connected) =>
```
Comment (reinierl) [inline src/main/scala/com/thenewmotion/akka/rabbitmq/ConnectionActor.scala]: Shouldn't you be setting / unsetting the `connectionBlocked` field in the state here?

**Comment by reinierl** [comment]
Regarding the application running out of memory because it's keeping all these messages in memory: that could happen. The solution would be to have a configurable maximum buffer size.

Still the existing behavior (queue messages until OOM) isn't that bad in our current use case, where having the producer shut down and not produce messages anymore is preferable to losing messages.

**Comment by reinierl** [comment]
Regarding confirmation of published messages: to be really sure that your message was published you need publisher confirms, as explained on https://www.rabbitmq.com/reliability.html#connection-failures. The fact that basicPublish() returned doesn't tell you enough yet.

**Comment by sebastian-alfers** [comment]
what is the status about this @reinierl @Morgaroth  ? would be helpful to be able to react on blocked connections

**Pull request closed without merging**